pcicrawler/ocp_diag.py

```diff
@@ -0,0 +1,259 @@
+        self._json_path = json_path
+        self._devs = devs
+        self._ocp_run = ocp_run
+        self._result = True
+
+    def run(self):
+        duts = []
+        # Process input json
+        with open(self._json_path, 'r') as fh:
+            input_json = json.load(fh)
+            try:
+                duts = input_json['duts']
+            except KeyError as _:
+                raise json.JSONDecodeError("ERROR: Cannot parse file {}, missing DUTs info.".format(self._json_path))
+        filtered_duts = []
+        for dut in duts:
+            identifier = dut['identifiers']
+            potential_duts = list(self._devs)
+            # Go look for this device and get the data
+            missing_devices = []
+            if 'address' in identifier:
+                if identifier['address'] not in potential_duts:
+                    missing_devices.append("No device was found with address {}.".format(identifier['address']))
+                potential_duts = [dut for dut in potential_duts if str(dut) == identifier['address']]
+            if 'vendor_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.vendor_id == identifier['vendor_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with vendor id {}.".format(identifier['vendor_id']))
+            if 'device_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.device_id == identifier['device_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with device id {}.".format(identifier['device_id']))
+            if len(potential_duts) != 0:
+                filtered_duts.append(TestSet(potential_duts, dut['validate']))
+
+        # Create the dut information for the test run
+        dut = tv.Dut(id="0", name=socket.gethostname())
+        devices_seen = []
+        for test_set in filtered_duts:
+            for pcie_device in test_set.duts:
+                if pcie_device not in devices_seen:
+                    devices_seen.append(pcie_device)
+                    dut.add_hardware_info(name=pcie_device.name,
+                                          location=str(pcie_device.location),
+                                          manufacturer=str(pcie_device.vendor_id),
+                                          )
+        dut.add_software_info(name="pcicrawler",
+                              revision=pkg_resources.get_distribution("pcicrawler").version)
+
+        # Start checks
+        with self._ocp_run.scope(dut=dut):
+            # First we can list out the missing devices
+            if len(missing_devices) > 0:
+                self._result = False
+                for missing_device_msg in missing_devices:
+                    self._ocp_run.add_log(tv.objects.LogSeverity.ERROR, missing_device_msg)
+            for test_set in filtered_duts:
```

Comment:
are these duts or test sets? the names dont match

Reply:
test set contains both the test to run and DUT information, just going to rename it to test_sets